### ISSUE TITLE
Benchmark is Incomplete and Misleading

### ISSUE THREAD

**Issue by ackava** [dotnet/EntityFramework.Docs#3994]
We have recently used TPH over TPT and we found that queries are very slow when TPH is used as owned types.

We already had used TPT in the past, we decided to migrate to TPH based on this performance, it seems the performance calculation is incomplete and misleading.

Problems with TPH
1. You need to have identity column.
2. Unique entities would require additional unique index with filter.
3. Joins with other tables will slow down as SQL server will skip indexes due to Discriminators.

The last point slows down SQL server by 100 folds, our queries that finishes in milliseconds with 100000 of items, it now takes 6-7 seconds with only 1000 items.

We didn't even get to go live.

Here is a sample model.

```c#

   class Job {

       public List<LocationConstraint> Locations { get; set; }
       public List<AgeConstraint> Ages { get; set; }
       public List<EducationConstraint> Educations { get;set; }
   }
  
   class Constraint {
        [Key, DatabaseGenerated.....]
        public long ConstraintID { get; set; }
        public long JobID {get;set;}
   }

   class LocationConstraint : Constraint {
        public long CityID {get;set;}       
        public double Radius {get;set;}
        [ForeignKey(nameof(CityID))]
        public City City { get; set; }
   }

   class AgeConstraint: Constraint {
        public int MinAge {get;set;}
        public int MaxAge {get;set;}
   }

   class EducationConstraint: Constraint {
        public int Level {get;set; }
   }

```

In this searching jobs with some constraints, matching by location, by age and by other degree, is extremely slow. The generated query is very complex for even simple search with two constraints.

In order to search jobs by location range, the join to City table through LocationConstraint becomes very slow.

One could argue that we shouldn't use this in terms of owned types, but this is just for reference.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: deb8ded7-f955-4e7e-c711-722af98c84d7
* Version Independent ID: 9acdd9fa-8c8f-7548-83bd-8ff966d78fc8
* Content: [Modeling for Performance - EF Core](https://docs.microsoft.com/en-us/ef/core/performance/modeling-for-performance#inheritance-mapping)
* Content Source: [entity-framework/core/performance/modeling-for-performance.md](https://github.com/dotnet/EntityFramework.Docs/blob/main/entity-framework/core/performance/modeling-for-performance.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @roji
* Microsoft Alias: **avickers**

**Comment by roji** [comment]
@ackava can you please provide more details on your scenario? I'm missing specifically what indexes you've defined, and which exact query you're running which is slow. A full, minimal code sample would help us make sure we're talking about the same thing.

In the meantime, some responses:

> You need to have identity column.

This point is very unclear to me - TPH doesn't require an identity column any more than TPT. Both TPT and TPH can have any key type; it can be an identity column, a GUID, or a manually-managed key which you insert each time. I don't see how this can negatively affect TPH performance in any way.

> Unique entities would require additional unique index with filter.

That really depends on exactly what you're trying to achieve... If you want some property to be unique across your entire hierarchy, then no additional unique index is required: you can just define a unique index on that property at the root entity level as usual. Can you please provide more context on exactly what you're trying to do here?

> Joins with other tables will slow down as SQL server will skip indexes due to Discriminators.

So stepping back... In some scenarios, TPH may require you to define your indexes with the discriminator in mind; that is a normal, expected requirement, and not a performance problem of TPH itself. Can you please provide the exact query and the indexes you're defining, so that we can understand what's going on and help?

**Comment by ackava** [comment]
@roji I agree for first two points, although little overhead but still there is an additional issue when half of the entities needs to be unique and other don't.

Regarding join, we found the following issue after investigating generated SQL and execution plan inside SQL Server.

SQL Server cannot use index based on the value of discriminator, we did create index involving discriminator as primary and covering indexes but still no benefit. Reason is, SQL Server does not have statistics of how many entity types exist, for example LocationConstraint may be 20% of total rows in Constraints table. In this case, SQL Server skips indexes, this is still ok as long as you are querying same table with single constraint, but when you perform join (along with other constraints) of LocationConstraint with Cities table , which contains 2 million rows, SQL server opts scanning primary indexes of both Constraints and Cities tables. This results in slow query.

Where else if LocationConstraint is a separate table, SQL Server uses correct index and query results in few milliseconds.

The benchmark is misleading because it does not include actual query scenario when derived tables have other joins and those joins are are part of query.

I believe SQL Server and any other RDBMS will not choose index based on discriminator value supplied. Unless if we create indexes with constant filter like 'Discriminator = 'LocationConstraint' in the index. I haven't tried but I still doubt this will work.

**Comment by roji** [comment]
> SQL Server cannot use index based on the value of discriminator [...]
> ...
> I believe SQL Server and any other RDBMS will not choose index based on discriminator value supplied.

A TPH discriminator is nothing more than an additional column on the table, which EF uses to distinguish between entity types. It's not different from any other additional column which you might add in your model, and which you'd want to index. SQL Server can absolutely use indexes which are based on the value of discriminators, but the index may need to be defined correctly to make sure that works. In other words, when migrating from TPT to TPH you may need to change the indexes to make things work efficiently; if you haven't done that, then TPH may appear "slow", but the problem is actually with your indexes.

> Reason is, SQL Server does not have statistics of how many entity types exist [...]

This doesn't sound right. SQL Server does maintain statistics on tables based on the data inside them, and chooses whether to use indexes or not based on those statistics. This is done because sometimes using an index can be slower than not using it.

In summary, if you're seeing a performance problem related to the discriminator, I strongly suspect there is something wrong with the way you're defining your indexes, and not any problem specific to TPH itself. What we need here is a concrete, minimal code sample which shows a minimal model and the indexes you've defined. That would allow me to investigate and help make sure things work efficiently.

**Comment by ackava** [comment]
@roji 
> This doesn't sound right. SQL Server does maintain statistics on tables based on the data inside them, and chooses whether to use indexes or not based on those statistics. This is done because sometimes using an index can be slower than not using it.

I forgot to add, SQL Server does maintain statistics if Discriminator column is the first column, but in our case, Index with (JobID and Discriminator). So I went ahead and created index with (Discriminator and JobID) and tried to see if SQL Server chooses this index. SQL Server still ignores this index as there are other filter criteria present.

I think I am close to narrow down the issue.

## The real issue is `OR` predicate.

```mermaid
classDiagram
Constraint <|-- TagConstraint
Constraint <|-- FileConstraint
Constraint <|-- DateConstraint
Constraint <|-- RemoteConstraint
TagConstraint: Tag
FileConstraint: FileName
FileConstraint: Folder
FileConstraint: Size
DateConstraint: StartDate
DateConstraint: EndDate
RemoteConstraint: Protocol
RemoteConstraint: Url

class Job {
     List~TagConstraint~ Tags;
     List~FileConstraint~ Files;
     List~DateConstraint~ Date;
     List~RemoteConstraint~ Urls;
}
```

Eventually for better performance, we will need to declare many indexes, one for each constraint (discriminator) type.

```sql
CREATE UNIQUE INDEX IX_Jobs_TagConstraints ON Constraints ( Discriminator, JobID, Tag ) WHERE Tag is not null;
CREATE UNIQUE INDEX IX_Jobs_TagConstraints ON Constraints ( Discriminator, JobID, Url ) WHERE Url is not null;
```

I have tried using indexes without WHERE filter as well, but still SQL Server does not use index.

Now, here is an interesting case, when we create a query using `OR` of two different entity types, for example, 
```c#
   var tags = ["a", "b"];
   var urlPrefix = "https://%";
   var q =  db.Jobs
       .Where(x => !x.Tags.Any() || x.Tags.Any(t => tags.Contains(t.Tag))
       .Where(x => !x.Urls.Any() || x.Urls.Any(u => EF.Functions.Like(u.Url, urlPrefix));
```

Basically, we want to match constraints only if they are defined.

If both constraints are on separate tables with corresponding indexes, SQL Server does use index correctly. And results are faster.

Our queries on our models are far complex, and I have been struggling to find ways to improve it, But whichever way we try to generate indexes, TPT is faster, 1000 times faster.

**Comment by roji** [comment]
@ackava in order to help, I'm still missing an actual runnable code sample here. Your indexes somehow mismatch your query, possibly because of the order of the columns in the index; but I can't do much with the partial information above. At the very least please post the full LINQ query (the above seems incomplete, there's no ToList or similar at the end), the SQL produced from it, the plan SQL Server creates for it ([this page](https://docs.microsoft.com/en-us/ef/core/performance/performance-diagnosis?tabs=simple-logging%2Cload-entities#inspecting-query-execution-plans) may help). But ideally you'd set up a minimal, runnable code sample that I can run and see the issue.

**Comment by ackava** [comment]
@roji It will take sometime before I can make actual runnable code sample, I just can't put our private code base, I will have to redesign similar model with queries, I will reopen when I have the code.
